src/win/i2c.cc

```diff
@@ -0,0 +1,268 @@
+
+  void Execute() {   
+    String^ deviceSelector = I2cDevice::GetDeviceSelector(ref new Platform::String(controllerName.c_str()));
+    DeviceInformationCollection^ i2cDeviceControllers = create_task(DeviceInformation::FindAllAsync(deviceSelector)).get();
+    
+    if (nullptr == i2cDeviceControllers) {
+      return Nan::ThrowError(Nan::ErrnoException(EPERM, "getController", "DeviceInformation::FindAllAsync failed to return controller(s)"));
+    }
+
+    obj->SetControllerId(i2cDeviceControllers->GetAt(0)->Id);
+  }
+
+  void HandleOKCallback() {
+    Nan::HandleScope scope;
+    v8::Local<v8::Value> argv[] = { Nan::Null() };
+    callback->Call(1, argv);
+  }
+
+private:
+  WinI2c* obj;
+  std::wstring controllerName;
+};
+
+NAN_METHOD(WinI2c::GetController) {
+  if (info.Length() < 1 ||
+      !info[0]->IsString() ||
+      !info[1]->IsFunction()) {
+    return Nan::ThrowError(Nan::ErrnoException(EINVAL, "getController",
+      "incorrect arguments passed to getController"
+      "(string i2cControllerName, function cb)"));
+  }
+
+  v8::String::Utf8Value str(info[0]->ToString());
+  std::string stdStr = std::string(*str);
+  std::wstring controllerName;
+  controllerName.assign(stdStr.begin(), stdStr.end());
+
+  Nan::Callback *callback = new Nan::Callback(info[1].As<v8::Function>());
+  WinI2c* obj = Nan::ObjectWrap::Unwrap<WinI2c>(info.This());
+  
+  Nan::AsyncQueueWorker(new GetControllerWorker(callback, obj, controllerName));
+}
+
+NAN_METHOD(WinI2c::GetControllerSync) {
+  if (info.Length() < 1 ||
+      !info[0]->IsString()) {
+    return Nan::ThrowError(Nan::ErrnoException(EINVAL, "getControllerSync",
+      "incorrect arguments passed to getControllerSync"
+      "(string i2cControllerName)"));
+  }
+
+  v8::String::Utf8Value str(info[0]->ToString());
+  std::string stdStr = std::string(*str);
+  std::wstring stdWStr;
+  stdWStr.assign(stdStr.begin(), stdStr.end());
+
+  String^ deviceSelector = I2cDevice::GetDeviceSelector(ref new Platform::String(stdWStr.c_str()));
+  DeviceInformationCollection^ i2cDeviceControllers = create_task(DeviceInformation::FindAllAsync(deviceSelector)).get();
+
+  if (nullptr == i2cDeviceControllers) {
+    return Nan::ThrowError(Nan::ErrnoException(EPERM, "getControllerSync", "DeviceInformation::FindAllAsync failed to return controller(s)"));
+  }
+
+  WinI2c* obj = Nan::ObjectWrap::Unwrap<WinI2c>(info.This());
+  obj->_i2cControllerId = i2cDeviceControllers->GetAt(0)->Id;
+}
+
+class CreateDeviceWorker : public I2cAsyncWorker {
+public:
+    CreateDeviceWorker(
+        Nan::Callback *callback,
+        WinI2c* obj,
+        I2cConnectionSettings^ settings
+        ) : I2cAsyncWorker(callback), obj(obj), settings(settings) { }
+
+    ~CreateDeviceWorker() {}
+
+    void Execute() {
+      I2cDevice^ device = create_task(I2cDevice::FromIdAsync(obj->GetControllerId(), settings)).get();
+      if (nullptr == device) {
+        return Nan::ThrowError(Nan::ErrnoException(EPERM, "createDevice",
```

Comment:
This isn't a good idea. The Execute method is not called on the event loop thread (the main thread.) It's called on a worker thread so it's important steer clear of everything related to NAN, uv, Node, and V8. The exception to this rule is uv_async_send().

The best thing to do at the moment is to replace `return Nan::ThrowError(...);` with
```cpp
SetErrorNo(EPERM);
SetErrorSyscall("createDevice");
return;
```
Unfortunately the nice error message gets lost. [I2cAsyncWorker](https://github.com/fivdi/i2c-bus/blob/master/src/util.h#L4-L64) could be extended to support the nice error message.